libcib, based, various: Minor fixes and improvements

nrwahl2 opened the pull request:
This addresses some things that I found while working on T185. Highlights:
* Deprecate unused `cib_api_operations_t` functions (see [T656](https://projects.clusterlabs.org/T656)).
    * I'm not tied to all of these, since some of them (particularly `bump_epoch()` and `erase()`) are potentially useful methods. They just don't have internal callers. We could keep them in order to provide a more complete API for external users.
    * @kgaillot said in T656: "My long-term goal is to deprecate all of libcib, replaced with equivalent functionality in libcrmcommon (probably a long ways off though)."
    * `ping()` could also be deprecated if it turns out that sbd doesn't need it; see comment https://projects.clusterlabs.org/T656#9046, etc.
* Deprecate CIB global callbacks. There are no internal users. It's confusing because nothing about the nomenclature suggests that it's a global callback, and nothing about the per-call callbacks' nomenclature suggests that those are per-call.
* Apply `cib_inhibit_notify` to replace notifications. It applies only to diff notifications currently, and that seems incorrect.
* Deprecate `cib_scope_local` and document its behavior. It's processed only when `pacemaker-based` is in legacy mode, so we can drop it when we drop support for mixed-version clusters with Pacemaker 1.1.11. This flag is misleading, because setting it outside of legacy mode does nothing -- it does **not** specify that only the local CIB gets modified.
* Replace `cib_operation_t` booleans with new enum flags for readability.
* Return `cib_operation_t` objects directly to callers.
    * Currently we return the object's index in an internal array, and then we pass that index to various getter functions.
    * With this PR, we access the object's members directly. The getters are short and don't save much code in the callers.
    * I'm not tied to this either. We could keep the getters and isolate `cib_operation_t` to one `based_common.c`.

I'm currently running 100 CTSlab tests. I ran 25 recently and everything was fine. I've only made a couple of small tweaks since then.

nrwahl2 commented:
> I'm currently running 100 CTSlab tests. I ran 25 recently and everything was fine. I've only made a couple of small tweaks since then.

No issues

kgaillot commented on lib/cib/cib_client.c:
```diff
@@ -647,7 +647,7 @@ cib_new_variant(void)
 
     new_cib->cmds->upgrade = cib_client_upgrade;
-    new_cib->cmds->bump_epoch = cib_client_bump_epoch;
+    new_cib->cmds->bump_epoch = cib_client_bump_epoch; // Deprecated method
```
I'm leaning to not deprecating this one, it's useful, and unused internally only because cibadmin directly calls cib_internal_op() (which external callers can't do)

nrwahl2 replied:
I feel the same way. I made commits for this and `erase` for completeness, to discuss them.

kgaillot commented on lib/cib/cib_client.c:
```diff
@@ -628,7 +628,7 @@ cib_new_variant(void)
     new_cib->cmds->register_callback_full = cib_client_register_callback_full;
 
-    new_cib->cmds->noop = cib_client_noop;
+    new_cib->cmds->noop = cib_client_noop; // Deprecated method
```
FYI we should keep ping; all daemons should implement a ping (and hopefully one day we can consolidate that code across daemons)

nrwahl2 replied:
Removing support for `CRM_OP_PING` was never under consideration. The only thing I was considering was deprecating the `ping()` method of `cib_api_operations_t`, since it's unused internally.

I'm fine with keeping it though. Like we discussed for `bump_epoch()` and `erase()`, they're potentially useful parts of a complete API, even if we don't use them as such internally.

kgaillot commented on lib/cib/cib_client.c:
```diff
@@ -654,7 +654,7 @@ cib_new_variant(void)
     new_cib->cmds->replace = cib_client_replace;
     new_cib->cmds->remove = cib_client_delete;
-    new_cib->cmds->erase = cib_client_erase;
+    new_cib->cmds->erase = cib_client_erase; // Deprecated method
```
This is another one where cibadmin calls cib_internal_op() directly -- probably useful to keep

nrwahl2 replied:
agreed

kgaillot commented on lib/common/xml.c:
```diff
@@ -686,12 +698,26 @@ create_xml_node(xmlNode * parent, const char *name)
     if (parent == NULL) {
         doc = xmlNewDoc((pcmkXmlStr) "1.0");
+        if (doc == NULL) {
+            return NULL;
```
We don't usually check create_xml_node() for NULL either, but that can be left for blue skies.

Presumably if we return NULL past this point, we have some freeing to do

I'm tempted to just assert on memory errors in create_xml_node(), even though I usually prefer returning an indication when possible

nrwahl2 replied:
I'm good either way. In the past my default approach was to assert on memory errors, like GLib does. You've said recently that you're moving away from doing that, since a core dump isn't useful.

I'd been meaning to ask you how your feelings translated to functions like `pcmk__str_update()`.

kgaillot replied:
My general feeling is that library functions in particular should bubble up errors since callers have more context to know the best reaction. But something like pcmk__str_update() is so widely used that checking the return value all the time would be a major hassle, both in initially converting it and in readability. So I don't like it but I accept it :)

nrwahl2 replied:
I'm leaving this alone for now; I'm not sure where we landed on changing it. Could do either:
* leave it alone for now, maybe make a wishlist task for NULL-checking `create_xml_node()`, or
* assert on memory error

kgaillot replied:
It's fine as is, except that we need to free any allocated memory before returning NULL

kgaillot commented on daemons/based/pacemaker-based.h:
```diff
@@ -43,10 +43,20 @@ enum cib_client_flags {
     const char *operation;
-    gboolean modifies_cib;
-    gboolean needs_privileges;
+    enum cib_op_attr flags;
```
uint32_t

nrwahl2 replied:
Ah yep. I can't remember *why* we prefer explicit types like `uint32_t` for a struct member... if it were just to ensure a strict size for the struct, we'd never use `int` either.

kgaillot replied:
Because the member does not hold an enum value, but a bitmask of enum values. It's unsigned since bitmasks should always be, and specifying the size ensures we have enough for the number of bits defined

nrwahl2 replied:
Ah, that's right.

kgaillot commented on daemons/based/based_callbacks.c:
```diff
@@ -579,24 +581,27 @@ parse_local_options_v2(const pcmk__client_t *cib_client, int call_type,
-                   gboolean * needs_forward)
+parse_peer_options_v1(const cib_operation_t *operation, xmlNode *request,
+                      gboolean *local_notify, gboolean *needs_reply,
+                      gboolean *process, gboolean * needs_forward)
```
missed one space :)

nrwahl2 replied:
thanks!

kgaillot commented on daemons/based/based_callbacks.c:
```diff
@@ -1277,7 +1289,11 @@ cib_process_command(xmlNode * request, xmlNode ** reply, xmlNode ** cib_diff, gb
+                                   PCMK__CIB_REQUEST_REPLACE,
+                                   NULL),
+                  crm_err("Call type: %s", operation->name);
                   crm_log_xml_err(request, "bad op"));
```
FYI we don't need the crm_err() or crm_log_xml_err() here; the CHECK logs failures (showing the condition tested), and this isn't a check we really care much about. But no need to change since it will all go away at some point anyway

nrwahl2 replied:
No need to *not* change it either, since there are other edits to make :)

kgaillot reviewed (commented):
Beautiful

nrwahl2 commented on daemons/based/pacemaker-based.h:
```diff
@@ -43,10 +43,20 @@ enum cib_client_flags {
+enum cib_op_attr {
+    cib_op_attr_none       = 0,         //!< No special attributes
+    cib_op_attr_modifies   = (1 << 1),  //!< Modifies CIB
+    cib_op_attr_privileged = (1 << 2),  //!< Requires privileges
```
The stickler in me doesn't like using a verb form in one and an adjective form in another (non-parallel structure). A future commit will introduce `cib_op_attr_transaction` (noun) though, so all bets are off. `*_supports_transaction` is too long, as is `*_needs_privileges`.

Also the meaning of `cib_op_transaction` (if we dropped "_attr") would be too easily confused.

kgaillot replied:
Agreed

nrwahl2 commented:
Addressed review, fixed a NULL check issue found by Coverity, and added three more cleanup commits on top. Hoping Coverity doesn't complain again about the `pcmk__file_contents()` false positive...

kgaillot commented on daemons/based/based_messages.c:
```diff
@@ -61,25 +61,15 @@ cib_process_shutdown_req(const char *op, int options, const char *section, xmlNo
-        crm_err("Action [%s] is not supported by the CIB manager", op);
-    }
-    return result;
+    return -EPROTONOSUPPORT;
```
This is what it previously did when it *wasn't* noop

nrwahl2 replied:
Crap, yeah. My brain automatically said "If there are two conditions and a default case, the default case won't be the `else if`."

kgaillot reviewed (commented):
looking good

nrwahl2 commented:
It's explicitly marked as a false positive and I haven't even touched this code.......
```
===== NEW COVERITY SCAN ERRORS DETECTED =====
Error: STRING_NULL (CWE-170):
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/cib/cib_file.c:578: string_null_source: Function "pcmk__file_contents" does not terminate string "expected".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/cib/cib_file.c:580: path: Switch case value "pcmk_rc_ok".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/cib/cib_file.c:582: path: Condition "expected == NULL", taking false branch.
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/cib/cib_file.c:586: path: Breaking from switch.
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/cib/cib_file.c:595: string_null: Passing unterminated string "expected" to "pcmk__verify_digest", which expects a null-terminated string.
===== END NEW COVERITY SCAN ERRORS DETECTED =====
```

nrwahl2 commented:
fixed the other stuff; not sure what to do about Coverity's BS

nrwahl2 commented:
The Coverity error starts with the "[NULL-check argument to pcmk__mark_xml_created()](https://github.com/ClusterLabs/pacemaker/pull/3105/commits/29228a5360ff06a5fcf9e5e634882f75dcdfa4c3)" commit.

kgaillot commented on lib/common/xml.c:
```diff
@@ -686,12 +699,28 @@ create_xml_node(xmlNode * parent, const char *name)
+
         node = xmlNewDocRawNode(doc, NULL, (pcmkXmlStr) name, NULL);
+        if (node == NULL) {
+            xmlFreeDoc(doc);
```
we don't want to free in this case, since parent remains valid

nrwahl2 replied:
Ooo, good catch.

kgaillot commented on daemons/based/pacemaker-based.h:
```diff
@@ -55,12 +55,11 @@ enum cib_op_attr {
-    const char *operation;
-    enum cib_op_attr flags;
+    const char *name;
+    uint32_t flags; //!< Group of <tt>enum cib_op_attr</tt> flags
```
FYI the type correction is in the cib_get_operation() commit instead of the booleans commit, but no big deal

nrwahl2 replied:
moved this while fixing the xmlFreeDoc issue

kgaillot reviewed (commented):
Looks good, one small thing

nrwahl2 commented on lib/common/io.c:
```diff
@@ -460,11 +460,17 @@ pcmk__file_contents(const char *filename, char **contents)
         rewind(fp);
-        read_len = fread(*contents, 1, length, fp); /* Coverity: False positive */
+
+        read_len = fread(*contents, 1, length, fp);
```
Moving the newline was an accident but it's not worth undoing now.
